### DISCUSSION TITLE
Raw memory passing

### DISCUSSION THREAD

**Discussion by Scooletz** [sebastienros/jint#1501] [Ideas]
I did my best to understand `JsValue` and its semantics of value passing. Has it been considered to provide another `JsValue` that could represent an `ArrayBufferInstance` that would be based on `ReadOnlyMemory<byte>`. This could be beneficial for scenarios where raw data are loaded in .NET runtime and passed to some script for executing logic over it. Additionally, depending on how closely Jint needs to be in regards to the buffer detachment, an `ArrayBufferInstance` that is based on `ReadOnlyMemory<byte>` could allow a clone-less copying, just by slicing over the underlying `ReadOnlyMemory<byte>`.

**Comment by lahma** [comment]
This sound interesting. So do you have an actual use case that you are building?

`ArrayBufferInstance` is by nature read-write memory so it doesn't directly map to ROM. How would you use this type, constructed on C# side and then fed to engine via `engine.SetValue("variable", myInstance)` or would this require deeper customization like that you would control in instantiation of default `ArrayBufferInstance`? `ArrayBufferInstance` could probably model it's backing store using `Memory<T>` (write required). Would your custom type be inheriting from default `ArrayBufferInstance` and overriding some methods and throwing `InvalidOperationException` in case of writes?

There's also [DataView](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/DataView) to consider.

**Comment by Scooletz** [comment]
I'm currently sketching a filtering engine that would either run the filter in .NET or in JS. As the data are uint32, which falls into `TypedArray` and `ArrayBuffer`. If the latter option is selected, it would be nice to be able to pass large slabs of data over without copying at all. I'm somewhat familiar with the `DataView` but as it's created on top of an `ArrayBuffer` and in Jint it's internal, I don't know how this helps or prohibits `ArrayBuffer` considerations.

If the idea seems plausible, maybe I can dive a bit deeper into Jint and try to sketch a PR with it :-)

**Reply by lahma** [reply]
A PR would be most welcome and appreciated! Like any OSS project, we happily accept help and appreciate it, there's never too much of help available. Many things in Jint are modeled based on JS spec but we would love to have first class .NET experience with comfortable APIs that "make things just work".

Some background:

* We make things internal at first to allow some wiggle room, when a plausible use case is found things can be made public and just require some more scrutiny to ensure we don't create "bad APIs", we have `Jint.Tests.PublicInterface` for the API promises we've made and try to maintain
* We go [spec](https://tc39.es/ecma262/) first, so EcmaScript has all sorts of weirds things that need to be supported, some code isn't optimal at first (hopefully compliant!), but we try to optimize when performance problems are found - as an example arrays in JS have a lot of special casing but we try to make them fast

With your track record, any PR to improve the project is most welcome 😉

**Reply by Scooletz** [reply]
Sounds like a plan. I'll dig deeper then and if in doubt will raise it here.

**Comment by Scooletz** [comment]
Will be implemented in PR #1590. Closing the discussion as the rest will be implementation specific.
